### PR TITLE
Login logout

### PR DESCRIPTION
**Pull request by leahriffell**
Closes login/logout milestone

**Contributors:**
Priya, Luke, Grant, Leah

**What this did**
This epic allows users to log in and log out.

### CODE REVIEW

File: app/assets/stylesheets/application.css [hunk 1]
```diff
@@ -128,6 +128,8 @@ form {
   background-color: #f2f2f2;
   padding: 20px;
   width: 60%;
+  margin-left: auto;
+  margin-right: auto;
```
Comment (LHJE) [inline app/assets/stylesheets/application.css:132]: Hmm, was this change something we did?  I don't recall this one, but maybe that was some solo work somewhere.  I'm sure it's fine, but I wanna double check.

File: app/controllers/merchants/base_controller.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class Merchants::BaseController < ApplicationController
+  before_action :require_merchant
+
+  def require_merchant
+    if !current_merchant?  
+      render file: "/public/404"
```
Comment (LHJE) [inline app/controllers/merchants/base_controller.rb:6]: Love that this worked.

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+class SessionsController < ApplicationController
+
+  def new
+    if current_user
+      # Can we make this condition a model method (maybe a case statement)? Use it to refactor sessions#create as well?
```
Comment (LHJE) [inline app/controllers/sessions_controller.rb:5]: Let's take out this commented info.

File: app/controllers/welcome_controller.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class WelcomeController <ApplicationController
+
+  def index
+
```
Comment (LHJE) [inline app/controllers/welcome_controller.rb:4]: Small, but we could take this extra line out.

File: spec/features/users/login_spec.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+RSpec.describe "Login Page", type: :feature do
+  describe "As a visitor" do
+    before :each do
+      @user = User.create(name:"Luke Hunter James-Erickson", address:"skdjfhdskjfh", city:"kajshd", state:"jsdh", zip:"88888", email: "tombroke@gmail.com", password:"Iamapassword", password_confirmation:"Iamapassword", role: 0)
+
+      @merchant = User.create(name:"Leah", address:"123 Sesame Street", city:"New York", state:"NY", zip:"90210", email: "Leahsocool@gmail.com", password:"Imeanit", password_confirmation:"Imeanit", role: 1)
+
+      @admin = User.create(name:"Priya", address:"13 Elm Street", city:"Denver", state:"CO", zip:"66666", email: "priyavcooltoo@gmail.com", password:"yuuuuuup", password_confirmation:"yuuuuuup", role: 2)
+
```
Comment (LHJE) [inline spec/features/users/login_spec.rb:11]: We can probably take out lines 9 and 7

File: spec/features/users/login_spec.rb [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    it "cannot redirect a logged in regular user to a dashboard" do
+      fill_in :email, with: @user.email
+      fill_in  :password, with: @user.password
+      click_button "Login"
+
+      visit "/merchants"
+        # visit "/merchants/dashboard"
+        # save_and_open_page
+        # Checked in a save and open page - some sort of 404 error is showing, we just need to use the PREBUILT errors
```
Comment (LHJE) [inline spec/features/users/login_spec.rb:137]: Let's get rid of these...

File: spec/features/users/login_spec.rb [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    #
+    #   visit '/merchants/dashboard'
+    #
+    #   expect(current_path).to eq("/public/404")
+    #
+    #   visit '/admin/dashboard'
+    #
+    #   expect(current_path).to eq("/public/404")
+    # end
+
```
Comment (LHJE) [inline spec/features/users/login_spec.rb:156]: ...and all of these.

File: spec/features/users/logout_spec.rb [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+      @merchant = User.create(name:"Leah", address:"123 Sesame Street", city:"New York", state:"NY", zip:"90210", email: "Leahsocool@gmail.com", password:"Imeanit", password_confirmation:"Imeanit", role: 1)
+
+      @admin = User.create(name:"Priya", address:"13 Elm Street", city:"Denver", state:"CO", zip:"66666", email: "priyavcooltoo@gmail.com", password:"yuuuuuup", password_confirmation:"yuuuuuup", role: 2)
+
+      @mike = Merchant.create(name: "Mike's Print Shop", address: '123 Paper Rd.', city: 'Denver', state: 'CO', zip: 80203)
+      @paper = @mike.items.create(name: "Lined Paper", description: "Great for writing on!", price: 20, image: "https://cdn.vertex42.com/WordTemplates/images/printable-lined-paper-wide-ruled.png", inventory: 25)
+      @pencil = @mike.items.create(name: "Yellow Pencil", description: "You can write on paper with it!", price: 2, image: "https://images-na.ssl-images-amazon.com/images/I/31BlVr01izL._SX425_.jpg", inventory: 100)
```
Comment (LHJE) [inline spec/features/users/logout_spec.rb:15]: We don't need line 15, and we can take out 8, 10, and 12

File: spec/features/users/logout_spec.rb [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+      expect(current_path).to eq("/")
+      expect(page).to have_content("Congrats, you're leaving, but you should stay!")
+      within(".topnav") do
+        expect(page).to have_link("Cart: 0")
+        click_link "Login"
+      end
+
+      expect(current_path).to eq('/login')
+    end
```
Comment (LHJE) [inline spec/features/users/logout_spec.rb:106]: I'm going to make an issue, if we haven't already, on drying this up during our refactor.  There's So much duplication, but I'm glad it's working properly.

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -11,13 +11,38 @@
...
   end
 
-  describe "class methods" do 
+  describe "class methods" do
     it ".email_exists?" do
-      user = User.create!(name:"Luke Hunter James-Erickson", address:"skdjfhdskjfh", city:"kajshd", state:"jsdh", zip:"88888", email: "tombroke@gmail.com", password:"Iamapassword", password_confirmation:"Iamapassword")
+      user = User.create!(name:"Luke Hunter James-Erickson", address:"skdjfhdskjfh", city:"kajshd", state:"jsdh", zip:"88888", email: "tombroke@gmail.com", password:"Iamapassword", password_confirmation:"Iamapassword", role: 0)
 
-      expect(User.email_exists?("tombroke@gmail.com")).to eq(true)
+      expect(User.email_exists?(user.email)).to eq(true)
```
Comment (LHJE) [inline spec/models/user_spec.rb:45]: Nice. More dynamic this way.

**Review by LHJE — CHANGES_REQUESTED** [review]
Just some minor cleanup that can be done before pushing it forward.

I'm going to make an issue for refactoring Logout_spec since there's so much duplication.

File: app/assets/stylesheets/application.css (hunk 1 above)
Comment (leahriffell) [reply app/assets/stylesheets/application.css:132]: Yup, it was a change from the last PR I did that Priya merged. I saw that the site is centering headers so the forms looked funky being left aligned.

File: app/controllers/merchants/base_controller.rb (hunk 2 above)
Comment (leahriffell) [reply app/controllers/merchants/base_controller.rb:6]: 💯 me too! In future stories we'll figure out how get these tests written properly.

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (leahriffell) [reply app/controllers/sessions_controller.rb]: Good catch. Done!

File: app/controllers/welcome_controller.rb (hunk 4 above)
Comment (leahriffell) [reply app/controllers/welcome_controller.rb:4]: Done!

File: spec/features/users/login_spec.rb (hunk 5 above)
Comment (leahriffell) [reply spec/features/users/login_spec.rb:9]: Donezo

File: spec/features/users/login_spec.rb (hunk 6 above)
Comment (leahriffell) [reply spec/features/users/login_spec.rb]: Done

File: spec/features/users/login_spec.rb (hunk 7 above)
Comment (leahriffell) [reply spec/features/users/login_spec.rb]: ...and done!

File: spec/features/users/logout_spec.rb (hunk 8 above)
Comment (leahriffell) [reply spec/features/users/logout_spec.rb]: Good catch on the unneeded variable! Removed that and the extra lines.

File: spec/features/users/logout_spec.rb (hunk 9 above)
Comment (leahriffell) [reply spec/features/users/logout_spec.rb:103]: Yeah good idea. Thanks for making an issue so we don't forget to revisit this!

File: app/assets/stylesheets/application.css (hunk 1 above)
Comment (LHJE) [reply app/assets/stylesheets/application.css:132]: Nice!

**Review by LHJE — APPROVED** [review]

**Pull request merged**